### ISSUE TITLE
Can't change CastleWindow size when open

### ISSUE THREAD

**Issue by eugeneloza** [castle-engine/castle-engine#36]
I can't seem to change the `TCastleWindowCustom` children `Window.height/width` runtime. I can easily set `Window.height/ Window.width` before `Window.Open`, but not after. `Window.doResize` also doesn't work. `Window.ResizeAllowed := raOnlyAtOpen` also doesn't prevent the user from resizing the Window in case it is assigned after `window.open`.
However, I can set Window size in Initialization right after creating the window instance and it'll work correctly before the window is open.
P.S. I'm absolutely sure it worked before, e.g. in April-May.
P.P.S. setting fullscreen to "true" normally enters fullscreen mode and resizes the Window.
Simple example attached.
[ResizeProblem.tar.gz](https://github.com/castle-engine/castle-engine/files/602900/ResizeProblem.tar.gz)

**Comment by michaliskambi** [comment]
It's simply not implemented. We do not (right now) react to some window changes after window is open, they are applied only at the next Window.Open. You have to assign them before calling Window.Open.

This applies to
- Width / Height 
- Left / Top
- ResizeAllowed 

It was like this since the beginning of CastleWindow existence, I'm 100% sure:) There is also a comment at Width / Height: "When the window is open, these are read-only..." in castlewindow.pas.

I have added now a SetXxx methods that make a warning if you try to change these properties while the window is open. If you have log initialized (or Application.OnWarning), you will see a clear warning in this case.

I have also changed DoResize to be private. It was made public only temporarily, for initial NPAPI code in https://github.com/castle-engine/castle-engine/commit/4be89ea604d687904cf2f8c74b11bdb48e5cd89c . It is no longer needed, as the NPAPI code is now in the same unit.

Contributions to improve this (make window actually react to Width / Height changes) are welcome. It should be easy, but will require different code for different backends. You will want to define SetWidth and friends in every backend.

Possibly, these changes should be then applied with some delay (e.g. in DoUpdate?), not instantly in SetWidth or SetHeight. You will often assign both Width and Height, and the size should be applied only once. I guess that LCL or VCL do it like that. Or maybe a public "Resize(NewWidth, NewHeight)" would be OK then (the GTK does it like that: https://developer.gnome.org/gtk3/stable/GtkWindow.html#gtk-window-resize ).

**Comment by eugeneloza** [comment]
> I'm 100% sure:) 

As strange as it might look I am sure it worked for me in FireMadness... I remember I've been testing it and it worked as expected in Windows (preventing the window from scalling too small... https://github.com/eugeneloza/FireMadness/blob/master/FireMadness.pas#L1558 I wouldn't have left this procedure if I didn't test it thoroughly at that moment... hmm..... 
But anyway, I must have been wrong then :)

I'll look in LCL implementation. But from the first glance looks too complex for me... :)

**Comment by michaliskambi** [comment]
It definitely never worked, I never implemented it:)

Also, "Window.DoRender" should not be public (was only temporarily made public), fixed now. You should call "Window.Invalidate" to force redraw at the nearest time, this will work OK always and on all platforms, and will keep features like Application.LimitFPS working OK.

To force a minimal window size, you can set Window.MinWidth and Window.MinHeight. (Before the Window.Open.)

> I'll look in LCL implementation. But from the first glance looks too complex for me... :)

You just need to be familiar with particular OS-specific APIs. E.g. the "WinAPI" backend, used by default on Windows, is mostly in src/window/windows/castlewindow_winapi.inc . The "GTK" backend, used by default on Linux (but cross-platform in theory), is mostly in src/window/gtk/castlewindow_gtk.inc .

Inside such include file, you could define a method like SetWidth that sets window size using the WinAPI or GTK function, like https://msdn.microsoft.com/en-us/library/ms633545(VS.85).aspx and https://developer.gnome.org/gtk2/stable/GtkWindow.html#gtk-window-resize

The backends code is relatively simple, it's just spread out across a couple of files. Also, some of this code is *really* old (when I started implementing the engine, the WinAPI and Xlib backends of CastleWindow were one of the first things that had to work...), so I'm not proud of it:) Making each backend a separate class, that descends from some TCastleWindowBackend class, would be nicer (and will eventually happen:).

But it should be manageable for now:)

**Comment by michaliskambi** [comment]
Oh, and indeed the LCL implementation (in src/window/castlewindow_lcl.inc ) is probably the simplest one. It implements CastleWindow using Lazarus LCL underneath.

But it's also not used by default on Windows (because "WinAPI" backend is better), or Linux (because "GTK" backend is better), or Android (because castlewindow_android.inc backend is better)... The castlewindow_lcl.inc is used by default only on Mac OS X now. But it should work everywhere (well, at least on all desktop platforms, not on mobile / web plugin).

**Comment by eugeneloza** [comment]
> Also, "Window.DoRender" should not be public 

Yes... that was my mistake. I already know about it, but haven't fixed it yet :)

> you can set Window.MinWidth and Window.MinHeight

Yes, I also discovered this cool feature lately. That will do perfectly for FireMaddness But in my new project (Decoherence) I need window size to have a specific ratio, not size. Still, I think I'll just prevent the user from rescaling the window :) At least at this development stage.

**Comment by michaliskambi** [comment]
>  Still, I think I'll just prevent the user from rescaling the window :) At least at this development stage.

An alternative is to allow for *any* window size, and only force the "actual game area" inside the window to have a specific aspect ratio. That is, calculate the "actual game area" size to fit within the window, fitting the width / height to window, and adjusting the other dimension to follow the aspect ratio. Around the "actual game area" you can simply render blackness (this can be done easily e.g. by "Window.Controls.InsertBack(TCastleSimpleBackground.Create(Application))").

The "actual game area" can be many things, depending on how you render. 

- If you render everything through a scene manager, you can just adjust the Left/Bottom/Width/Height of your TCastleSceneManager instance, and it will magically work.

- If you render yourself (in TUIControl.Render override, or Window.OnRender callback), then you have to adjust to the calculated rectangle yourself.  You can use RenderContext.ScissorBegin / RenderContext.ScissorEnd around your rendering, to cut the contents to the given rect.

While this is more work, the advantage is that you can work on Android, or in any fullscreen resolution. So it's more future-proof:) This is what I'm using in games that require specific aspect ratio, like "Dragon Squash".

**Comment by eugeneloza** [comment]
I'll also move this issue to my repository branch.
